### PR TITLE
p2p Operation Optimizing file writing

### PR DESCRIPTION
**Pull request by luyikang-York**

### CODE REVIEW

**Review by cyjseagull — APPROVED** [review]

File: libp2p/Service.cpp [hunk 1]
```diff
@@ -1193,20 +1195,30 @@ bool Service::updatePeersToIni(std::map<dev::network::NodeIPEndpoint, NodeID> co
         auto pos1 = fileData.find("p2p");
         if (pos1 == fileData.npos)
         {
+            SERVICE_LOG(DEBUG) << LOG_DESC("fail to locate peers' information in config file");
```
Comment (ywy2090) [inline libp2p/Service.cpp:1198]: SERVICE_LOG(DEBUG) => SERVICE_LOG(WARNING)

File: libp2p/Service.cpp [hunk 2]
```diff
@@ -1193,20 +1195,30 @@ bool Service::updatePeersToIni(std::map<dev::network::NodeIPEndpoint, NodeID> co
...
             return false;
         }
         auto pos2 = fileData.find("node.0=", pos1);
         auto pos3 = fileData.find("certificate_blacklist");
         if (pos2 == fileData.npos || pos3 == fileData.npos || pos2 >= pos3)
         {
+            SERVICE_LOG(DEBUG) << LOG_DESC("fail to locate peers' information in config file");
             return false;
         }
         fileData.replace(pos2, pos3 - pos2 - 1, tmpdata);
```
Comment (ywy2090) [inline libp2p/Service.cpp:1208]: the method to update nodeX fields is not appropriate， if the format of config.ini changed, eg: some other config 
items add before  certificate_blacklist, the replace operation will overide all the config items between p2p and certificate_blacklist

File: libp2p/Service.cpp [hunk 3]
```diff
@@ -1193,20 +1195,30 @@ bool Service::updatePeersToIni(std::map<dev::network::NodeIPEndpoint, NodeID> co
...
 
-        // TODO: 写入保护
-        std::ofstream out(confdir, std::ios::out);
-        out << fileData;
-        out.close();
+        auto newFileDir = confdir + ".new";
+        std::ofstream outfile(newFileDir, std::ios::out);
```
Comment (ywy2090) [inline libp2p/Service.cpp:1211]: add exception handle for io operation

File: libp2p/Service.cpp [hunk 4]
```diff
@@ -1193,20 +1195,30 @@ bool Service::updatePeersToIni(std::map<dev::network::NodeIPEndpoint, NodeID> co
...
+        outfile.flush();
+        outfile << fileData;
+        outfile.close();
+
+        if (0 != rename(newFileDir.c_str(), confdir.c_str()))
```
Comment (ywy2090) [inline libp2p/Service.cpp:1216]: if the rename operation failed, it means that the old config.ini maybe remove and the exist config is the new config.ini, and when the node restart, you should check this situation.

File: libp2p/Service.cpp [hunk 5]
```diff
@@ -1193,20 +1195,30 @@ bool Service::updatePeersToIni(std::map<dev::network::NodeIPEndpoint, NodeID> co
...
+        {
+            SERVICE_LOG(DEBUG) << LOG_DESC("fail to update config file") << LOG_KV("path", confdir);
```
Comment (ywy2090) [inline libp2p/Service.cpp:1218]: SERVICE_LOG(DEBUG) =》 SERVICE_LOG()

**Review by ywy2090 — CHANGES_REQUESTED** [review]

File: libp2p/Service.cpp (hunk 4 above)
Comment (luyikang-York) [reply libp2p/Service.cpp:1216]: [https://pubs.opengroup.org/onlinepubs/9699919799/functions/rename.html](url)says “If the rename() function fails for any reason other than [EIO], any file named by new shall be unaffected.”

Comment (luyikang-York) [reply libp2p/Service.cpp:1216]: i wonder if it is save enough to use rename

File: libp2p/Service.cpp (hunk 2 above)
Comment (luyikang-York) [reply libp2p/Service.cpp:1208]: got it

**Review by ywy2090 — APPROVED** [review]

**Pull request merged**